### PR TITLE
[SCB-85] Upgrade Metrics File Output from 0.5.0 to 1.0.0-m1

### PR DESCRIPTION
**Pull request by zhengyangyong**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
foundation-metrics in version 0.5.0 had function write metrics into file,in new version 1.0.0-m1 this function had rebuild :
1. new module in metrics-extension : metrics-write-file,metrics-write-file-config,metrics-write-file-config-log4j,metrics-write-file-config-log4j2
2. new samples : metrics-write-file-log4j,metrics-write-file-log4j2,metrics-write-file-log4j-springboot,metrics-write-file-log4j2-springboot

Help file is here : https://github.com/zhengyangyong/incubator-servicecomb-website/blob/asf-site/_users/cn/metrics-write-file-extension-and-sample-in-1.0.0-m1.md

这个PR的内容是将0.5.0版本写Metrics文件的功能升级到1.0.0-m1，以及包含如何使用的Samples
老版本的代码计划将在1.0.0-m2中删除，m1版本新旧功能并存

### CODE REVIEW

File: metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>io.servicecomb</groupId>
+      <artifactId>metrics-write-file-config</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-api</artifactId>
+      <version>2.8.2</version>
```
Comment (WillemJiang) [inline metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml]: why do you specify the version here?  We have the BOM to manage the dependencies

File: metrics/metrics-extension/metrics-write-file-config-log4j2/src/main/java/io/servicecomb/metrics/extension/writefile/config/Log4J2MetricsFileWriter.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  private RollingFileAppender initLogger(String loggerName, String filePrefix) {
+    String fileName = Paths.get(rootPath, filePrefix + loggerName + ".dat").toString();
+    String filePattern = Paths.get(rootPath, filePrefix + loggerName + "-%i.dat").toString();
+
+    PatternLayout layout = PatternLayout.newBuilder().withPattern(PatternLayout.DEFAULT_CONVERSION_PATTERN).build();
+    SizeBasedTriggeringPolicy policy = SizeBasedTriggeringPolicy.createPolicy(maxFileSize);
+    DefaultRolloverStrategy strategy = DefaultRolloverStrategy.createStrategy(String.valueOf(maxFileCount), null, null,
+        null, null, false, config);
+
+    //TODO:in version 2.8 + , log4j2 request use RollingFileAppender.newBuilder,but throw No such static method exception
```
Comment (WillemJiang) [inline metrics/metrics-extension/metrics-write-file-config-log4j2/src/main/java/io/servicecomb/metrics/extension/writefile/config/Log4J2MetricsFileWriter.java]: What's you point?

File: metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml (hunk 1 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml]: Ok,moved

File: metrics/metrics-extension/metrics-write-file-config-log4j2/src/main/java/io/servicecomb/metrics/extension/writefile/config/Log4J2MetricsFileWriter.java (hunk 2 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file-config-log4j2/src/main/java/io/servicecomb/metrics/extension/writefile/config/Log4J2MetricsFileWriter.java]: will improve later,in 1.0.0-m2

**Comment by zhengyangyong** [comment]
Rebase on upstream master

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    return pickedMetrics;
+  }
+
+  private double round(double value, int places) {
+    if (!Double.isNaN(value) && !Double.isInfinite(value)) {
+      BigDecimal decimal = new BigDecimal(value);
+      return decimal.setScale(places, RoundingMode.HALF_UP).doubleValue();
+    } else {
+      return 0;
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java]: OK

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentFormatter.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Autowired
+  public SimpleFileContentFormatter() {
+    hostName = NetUtils.getHostName();
+    if (StringUtils.isEmpty(hostName)) {
+      hostName = NetUtils.getHostAddress();
+    }
+
+    //may any problem ?
+    if (RegistryUtils.getServiceRegistry() == null) {
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentFormatter.java]: RegistryUtils is a static class an init in CseApplicationListener,after bean init,so we need wait until it completed

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentFormatter.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Override
+  public Map<String, String> format(Map<String, String> input) {
+    return input.entrySet().stream()
+        .collect(Collectors.toMap(Entry::getKey, entry -> {
+          try {
+            return JsonUtils
+                .writeValueAsString(new OutputJsonObject(applicationName, hostName, entry.getKey(), entry.getValue()));
+          } catch (JsonProcessingException e) {
+            logger.error("error convert metrics data to json convert", e);
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentFormatter.java]: Oh,sorry

File: samples/metrics-write-file-sample/metrics-write-file-log4j-springboot/pom.xml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  <parent>
+    <artifactId>metrics-write-file-sample</artifactId>
+    <groupId>io.servicecomb.samples</groupId>
+    <version>0.6.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>metrics-write-file-log4j-springboot</artifactId>
+
+  <!--need exclusion log4j-over-slf4j-->
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply samples/metrics-write-file-sample/metrics-write-file-log4j-springboot/pom.xml]: Done

File: samples/metrics-write-file-sample/metrics-write-file-log4j/pom.xml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      <groupId>io.servicecomb.samples</groupId>
+      <artifactId>metrics-write-file-common</artifactId>
+    </dependency>
+
+    <!--metrics dependency-->
+    <dependency>
+      <groupId>io.servicecomb</groupId>
+      <artifactId>metrics-write-file-config</artifactId>
+    </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply samples/metrics-write-file-sample/metrics-write-file-log4j/pom.xml]: Done,dependency relation had changed

File: foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/MetricsServoRegistry.java [hunk 8]
```diff
@@ -563,7 +563,7 @@ public String getValue() {
   }
 
   private double round(double value, int places) {
-    if (!Double.isNaN(value) && !Double.isInfinite(value)) {
+    if (!Double.isNaN(value)) {
       BigDecimal decimal = new BigDecimal(value);
       return decimal.setScale(places, RoundingMode.HALF_UP).doubleValue();
     } else {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/MetricsServoRegistry.java]: already return, do not write "else"

File: metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  <dependencies>
+    <dependency>
+      <groupId>io.servicecomb</groupId>
+      <artifactId>metrics-write-file</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-api</artifactId>
```
Comment (wujimin) [inline metrics/metrics-extension/metrics-write-file-config-log4j2/pom.xml]: it seems that, this is not log4j2?

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+          round(metric.getValue().doubleValue(), doubleRoundPlaces)));
+    }
+    return pickedMetrics;
+  }
+
+  private double round(double value, int places) {
+    if (!Double.isNaN(value)) {
+      BigDecimal decimal = new BigDecimal(value);
+      return decimal.setScale(places, RoundingMode.HALF_UP).doubleValue();
+    } else {
```
Comment (wujimin) [inline metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java]: no "else"

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+import io.servicecomb.metrics.common.RegistryMetric;
+import io.servicecomb.metrics.core.publish.DataSource;
+import io.servicecomb.metrics.extension.writefile.config.MetricsFileWriter;
+import io.servicecomb.serviceregistry.RegistryUtils;
+import io.servicecomb.serviceregistry.api.registry.Microservice;
+
+@Component
+public class WriteFileInitializer {
+
+  private static final String METRICS_WINDOW_TIME = "servicecomb.metrics.window_time";
```
Comment (wujimin) [inline metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: seems that we already have this const in some place?

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  private void init() {
+    final Runnable poller = this::run;
+    Executors.newScheduledThreadPool(1)
+        .scheduleWithFixedDelay(poller, 0, metricPoll, MILLISECONDS);
+  }
+
+  public void run() {
+    //wait RegistryUtils init completed
```
Comment (wujimin) [inline metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: maybe bootListener is a better choice?

File: metrics/metrics-extension/pom.xml [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      <dependency>
+        <groupId>io.servicecomb</groupId>
+        <artifactId>metrics-write-file</artifactId>
+        <version>0.6.0-SNAPSHOT</version>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <dependencies>
+    <dependency>
```
Comment (wujimin) [inline metrics/metrics-extension/pom.xml]: only need to write core? because core depend on common?

File: metrics/metrics-extension/pom.xml [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    <groupId>io.servicecomb</groupId>
+    <version>0.6.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>metrics-extension</artifactId>
+  <packaging>pom</packaging>
+
+  <modules>
+    <module>metrics-write-file-config</module>
```
Comment (wujimin) [inline metrics/metrics-extension/pom.xml]: why need so many logic?
we just need to print the log.

if someone need metrics log to a new file, then do it in his log configuration.

File: metrics/metrics-sample/pom.xml [hunk 15]
```diff
@@ -32,5 +32,17 @@
   <modules>
     <module>metrics-performance-test</module>
   </modules>
-  
+
+  <dependencies>
+    <dependency>
+      <groupId>io.servicecomb</groupId>
```
Comment (wujimin) [inline metrics/metrics-sample/pom.xml]: choice one of them.

File: samples/metrics-write-file-sample/metrics-write-file-common/pom.xml [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    <artifactId>metrics-write-file-sample</artifactId>
+    <groupId>io.servicecomb.samples</groupId>
+    <version>0.6.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>metrics-write-file-common</artifactId>
+  <dependencies>
+    <dependency>
+      <groupId>io.servicecomb</groupId>
```
Comment (wujimin) [inline samples/metrics-write-file-sample/metrics-write-file-common/pom.xml]: DO NOT use rest-common + web, i remember told you more than one time.

File: samples/metrics-write-file-sample/metrics-write-file-log4j/src/main/resources/META-INF/spring/pojo.provider.bean.xml [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  ~ limitations under the License.
+  -->
+
+<beans xmlns="http://www.springframework.org/schema/beans"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns:context="http://www.springframework.org/schema/context"
+  xsi:schemaLocation="http://www.springframework.org/schema/beans classpath:org/springframework/beans/factory/xml/spring-beans-3.0.xsd
+		http://www.springframework.org/schema/context http://www.springframework.org/schema/context/spring-context-3.0.xsd">
+
+  <context:component-scan base-package="io.servicecomb.samples.mwf"/>
```
Comment (wujimin) [inline samples/metrics-write-file-sample/metrics-write-file-log4j/src/main/resources/META-INF/spring/pojo.provider.bean.xml]: create a new issue?
to find a way to add main class package to component-scan path automatically

File: samples/metrics-write-file-sample/metrics-write-file-log4j2-springboot/src/main/resources/microservice.yaml [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+  handler:
+    chain:
+      Provider:
+        default: bizkeeper-provider
+
+servicecomb:
+  metrics:
+    #metrics double value round places,default value is 1
+    round_places: 1
+    file:
```
Comment (wujimin) [inline samples/metrics-write-file-sample/metrics-write-file-log4j2-springboot/src/main/resources/microservice.yaml]: i really do think we need this mechanism
likes invented a new wheel.

File: samples/springmvc-sample/springmvc-provider/src/main/java/io/servicecomb/samples/springmvc/provider/SpringmvcHelloImpl.java [hunk 19]
```diff
@@ -73,7 +74,7 @@ public String metricsForTest() {
     try {
```
Comment (wujimin) [inline samples/springmvc-sample/springmvc-provider/src/main/java/io/servicecomb/samples/springmvc/provider/SpringmvcHelloImpl.java]: why not return Map<String, String>

each time you write JsonUtils.writeValueAsString to be a schema operation result, you need think if you have make something wrong.

File: foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/MetricsServoRegistry.java (hunk 8 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/MetricsServoRegistry.java]: Done

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java (hunk 10 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/SimpleFileContentConvertor.java]: Done

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java (hunk 12 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: my no need,this way is simple and enough

File: metrics/metrics-extension/pom.xml (hunk 13 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/pom.xml]: Done

File: metrics/metrics-extension/pom.xml (hunk 14 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/pom.xml]: 新版本要求输出Operation级别的日志（单Operation就会有15+，如果区分维度会有40+），日志文件数量急剧膨胀，手动配置log工作量会非常庞大，需要自动配置

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java (hunk 11 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: Yes,Create a new Class MetricsConfig contain this config key

File: samples/metrics-write-file-sample/metrics-write-file-common/pom.xml (hunk 16 above)
Comment (zhengyangyong) [reply samples/metrics-write-file-sample/metrics-write-file-common/pom.xml]: Done，sorry,this code is written more early...

File: samples/metrics-write-file-sample/metrics-write-file-log4j2-springboot/src/main/resources/microservice.yaml (hunk 18 above)
Comment (zhengyangyong) [reply samples/metrics-write-file-sample/metrics-write-file-log4j2-springboot/src/main/resources/microservice.yaml]: this config had applied in 0.5.0,we may improve it later

File: samples/springmvc-sample/springmvc-provider/src/main/java/io/servicecomb/samples/springmvc/provider/SpringmvcHelloImpl.java (hunk 19 above)
Comment (zhengyangyong) [reply samples/springmvc-sample/springmvc-provider/src/main/java/io/servicecomb/samples/springmvc/provider/SpringmvcHelloImpl.java]: Done,we don't need this code and forget delete;(IT test code had remove to demo-springmvc)

File: metrics/metrics-sample/pom.xml (hunk 15 above)
Comment (zhengyangyong) [reply metrics/metrics-sample/pom.xml]: Done

File: metrics/metrics-extension/pom.xml (hunk 14 above)
Comment (wujimin) [reply metrics/metrics-extension/pom.xml]: log line has no relationship with file
only care for which class write the log.

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+    if (RegistryUtils.getServiceRegistry() != null) {
```
Comment (wujimin) [inline metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: it seems all these "if" can be init logic, no need to run every time.

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java (hunk 12 above)
Comment (wujimin) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: ......ok

File: samples/metrics-write-file-sample/metrics-write-file-log4j/src/main/resources/META-INF/spring/pojo.provider.bean.xml (hunk 17 above)
Comment (zhengyangyong) [reply samples/metrics-write-file-sample/metrics-write-file-log4j/src/main/resources/META-INF/spring/pojo.provider.bean.xml]: https://issues.apache.org/jira/browse/SCB-194

File: metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java (hunk 20 above)
Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: 我发现在Bean初始化的时候，RegistryUtils还没有init，但我不想自己贸然去调用Init方法；由于写文件是定期（低频率）操作，每次检查应该也没问题；微服务完全启动后再开始输出数据也是合乎逻辑的

Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: Done

Comment (zhengyangyong) [reply metrics/metrics-extension/metrics-write-file/src/main/java/io/servicecomb/metrics/extension/writefile/WriteFileInitializer.java]: use boot listener listen EventType.BEFORE_REGISTRY for startOutput

**Pull request merged**